internal/hcsoci/pnp.go

```diff
@@ -0,0 +1,58 @@
+// +build windows
+
+package hcsoci
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/shimdiag"
+	"github.com/Microsoft/hcsshim/internal/uvm"
+	"github.com/pkg/errors"
+)
+
+const uvmPnpExePath = "C:\\Windows\\System32\\pnputil.exe"
+
+// createPnPInstallDriverCommand creates a pnputil command to add and install drivers
+// present in `driverUVMPath` and all subdirectories.
+func createPnPInstallDriverCommand(driverUVMPath string) []string {
+	dirFormatted := fmt.Sprintf("%s/*.inf", driverUVMPath)
+	args := []string{
+		"cmd",
+		"/c",
+		uvmPnpExePath,
+		"/add-driver",
+		dirFormatted,
+		"/subdirs",
+		"/install",
+		"&",
+	}
+	return args
+}
+
+// createPnPInstallAllDriverArgs creates a single command to add and install
+// a set of driver locations in the UVM
+func createPnPInstallAllDriverArgs(driverUVMDirs []string) []string {
+	var result []string
+	for _, d := range driverUVMDirs {
+		driverArgs := createPnPInstallDriverCommand(d)
+		result = append(result, driverArgs...)
+	}
+	return result
+}
+
+// execPnPInstallAllDrivers makes the call to exec in the uvm the pnp command
+// that installs all drivers previously mounted into the uvm.
+func execPnPInstallAllDrivers(ctx context.Context, vm *uvm.UtilityVM, driverDirs []string) error {
+	args := createPnPInstallAllDriverArgs(driverDirs)
+	req := &shimdiag.ExecProcessRequest{
+		Args: args,
+	}
+	exitCode, err := ExecInUvm(ctx, vm, req)
+	if err != nil {
+		return errors.Wrapf(err, "failed to install drivers in uvm with exit code %d", exitCode)
```

Comment:
Does pnputil give useful error output? It seems like we should be logging that somewhere.

Reply:
that's a good point. I don't really think that the error message is that useful, but there should be some log files on the system that contain more information. Maybe we could dump those.